src/engine.rs

```diff
@@ -93,6 +93,7 @@ pub const KEYWORD_FN_PTR: &str = "Fn";
 pub const KEYWORD_FN_PTR_CALL: &str = "call";
 pub const KEYWORD_FN_PTR_CURRY: &str = "curry";
 pub const KEYWORD_SHARED: &str = "shared";
+pub const KEYWORD_TAKE: &str = "take";
```

Comment:
We're creating a pair of new keywords here: `shared` and `take`.  Is it strictly necessary?  Not a big deal, but these new keywords do add up...

Reply:
@schungx Well, there are rare cases when the user might want to get a copy of inner data of shared(in other words to call `Dynamic::read::<Dynamic>()`(that I'm going to rename). In other words to turn shared to non-shared. An example of such case is presented in this test: https://github.com/schungx/rhai/pull/20/commits/ca64668e583ec7b34377a1db5a11294c5c2570ff#diff-378246b680ad7fdabf490dd2eb856754R314 Here the user calls registered function that borrows one argument by mutable reference, and another by value(but it equals to borrowing by reference too) which leads either to borrow checking error or to a deadlock. It's almost impossible to resolve it on the `Engine` level since the Engine is eagerly borrowing all parts of expression beforehand.

Reply:
Yes, that's right.  By having multiple references to the same piece of lock-protected data, you run the risk of a Rhai script trying to access something that is already locked somewhere else during the same call.  In sync, you get a deadlock.  In non-sync, you get a panic.

```rust
let x = shared(1);
x.some_func( { x = 42; x }, { x = 0; x } );
```

In this example, you depend first on the evaluation order of arguments, then there are technically three references to `x`, one mutable (the object), and two immutable (result of both arguments).

Reply:
@schungx Also worth to mention that on Linux RwLock may panic too if OS sends a `EDEADLK` signal.

Both situations of panic could be fixed by using `RwLock::try_*`/`RefCell::try_` methods that would return Result::Err instead of panic. Do you think we should fix it in this PR? I would live it for future improvements.

Reply:
I think O/S-level failures we can leave to later (probably introduce a new Error variant just for it).

However, deadlocks and panics due to borrow_mut after borrow are, IMHO, not acceptable.  Users will expect the Engine to take care that they don't happen, or has a way to trap it out.

Reply:
@schungx I agree. Let me try one thing, I believe I can resolve it in easy way.